### PR TITLE
Create single NuPkg for the entire compiler toolset

### LINKED ISSUE #33219: Better experience for using compiler via nupkg and alternating between core and full msbuild
Currently, there are two different compiler packages for full-framework msbuild and core msbuild, respectively.

However, consider that is common to work with both msbuilds in tandem. For example, alt-tab between command line where you use `dotnet` and VS.

The following seems like it would work:

``` xml
<PackageReference Include="Microsoft.Net.Compilers" Version="2.10.0" Condition="'$(MSBuildRuntimeType)' != 'Core'" />
<PackageReference Include="Microsoft.NETCore.Compilers" Version="2.10.0" Condition="'$(MSBuildRuntimeType)' == 'Core'"  />
```

But it doesn't quite. The problem is that the package used is determined at restore time and you can restore with one msbuild (e.g. `dotnet restore`) and build with another (e.g. `msbuild`)

When you do that, you get complicated errors like this:

```
C:\Users\nicholg\.nuget\packages\microsoft.netcore.compilers\2.10.0\tools\Microsoft.CSharp.Core.targets(52,5): 
error MSB4062: The "Microsoft.CodeAnalysis.BuildTasks.Csc" task could not be loaded from the assembly
C:\Users\nicholg\.nuget\packages\microsoft.netcore.compilers\2.10.0\build\..\tools\Microsoft.Build.Tasks.CodeAnalysis.dll. 
Could not load file or assembly 'System.Runtime, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or 
one of its dependencies. The system cannot find the file specified. Confirm that the <UsingTask> declaration is correct, that 
the assembly and all its dependencies are available, and that the task contains a public class that implements 
Microsoft.Build.Framework.ITask.
```

Working around this is possible, but hacky. See https://github.com/dotnet/corefx/pull/35134/commits/9c7bf6731325e3d79af1b83742496173d840633d

I suggest one the following:

1. Make a single package that just works on both msbuilds.
2. Make the full-framework and core packages no-op on mismatched msbuilds. This way they can both be referenced without conditions. 
3. (2) + a meta-package that references both and behaves like (1) would.

cc @ericstj @tannergooding @agocke @jaredpar

### PR DESCRIPTION
**Pull request by jaredpar**
This creates a NuPgk that contains both the desktop and coreclr compiler toolsets. Consumers can now use this single package instead of having to condition between the two versions. 

In the long term we will delete the desktop / coreclr specific packages For the time being though we will keep producing them to give teams a period to transition off of them. 

closes #33219

### CODE REVIEW

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Microsoft.Net.Compilers.Toolset.Package.csproj [hunk 1]
```diff
@@ -0,0 +1,59 @@
+﻿<!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
```
Comment (jaredpar) [inline src/NuGet/Microsoft.Net.Compilers.Toolset/Microsoft.Net.Compilers.Toolset.Package.csproj]: @dsplaisted, @nguerrera can you all take a look  at this file? It's the multi-TFM NuGet tool project. It functions on my machine but I'm sure I've done unnecessarily evil things here.

**Comment by jaredpar** [comment]
@agocke test. Did you get this? Want to see if an explicit @ on a draft PR still sends email.

**Comment by jaredpar** [comment]
@dotnet/roslyn-infrastructure for review

File: Roslyn.sln [hunk 2]
```diff
@@ -1,6 +1,6 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.28620.130
+# Visual Studio 15
```
Comment (jcouv) [inline Roslyn.sln]: Is this intentional?

File: src/Setup/DevDivVsix/CompilersPackage/Microsoft.CodeAnalysis.Compilers.Setup.csproj [hunk 3]
```diff
@@ -200,5 +200,5 @@ folder InstallDir:\MSBuild\15.0\Bin\Roslyn\zh-Hant
     </ItemGroup>
   </Target>
 
-  <Import Project="..\..\..\NuGet\Microsoft.Net.Compilers\CompilerArtifacts.targets"/>
-</Project>
\ No newline at end of file
+  <Import Project="..\..\..\NuGet\Microsoft.Net.Compilers.Toolset\DesktopCompilerArtifacts.targets"/>
+</Project>
```
Comment (jcouv) [inline src/Setup/DevDivVsix/CompilersPackage/Microsoft.CodeAnalysis.Compilers.Setup.csproj]: Related to this PR: could you add an entry to https://github.com/dotnet/roslyn/wiki/NuGet-packages to list this new package with a one-line description of the purpose?

File: src/NuGet/Microsoft.NETCore.Compilers/tools/bincore/RunCsc [hunk 4]
```diff
@@ -1,14 +0,0 @@
-#!/bin/sh
```
Comment (jcouv) [inline src/NuGet/Microsoft.NETCore.Compilers/tools/bincore/RunCsc]: Were this file and related ones below included in packages before? If so, does removing them risk breaking some consumers of those packages?

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Microsoft.Net.Compilers.Toolset.Package.csproj [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    <IsPackable>true</IsPackable>
+    <NuspecPackageId>Microsoft.Net.Compilers.Toolset</NuspecPackageId>
+    <IncludeBuildOutput>false</IncludeBuildOutput>
+    <DevelopmentDependency>true</DevelopmentDependency>
+    <PackageDescription>
+      .NET Compilers package.
+      Referencing this package will cause the project to be built using the specific version of the C# and Visual Basic compilers contained in the package, as opposed to any system installed version.
+
+      This package can be used to compile code targeting any platform, but can only be run using the desktop .NET 4.7.2+ Full Framework.
```
Comment (jcouv) [inline src/NuGet/Microsoft.Net.Compilers.Toolset/Microsoft.Net.Compilers.Toolset.Package.csproj]: >  but can only be run using the desktop .NET 4.7.2+ Full Framework.

Is that correct?

File: src/NuGet/Microsoft.Net.Compilers.Toolset/buildMultiTargeting/Microsoft.Net.Compilers.Toolset.props [hunk 6]
```diff
@@ -0,0 +1,4 @@
+﻿<!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
+<Project>
+  <Import Project="..\build\Microsoft.Net.Compilers.Toolset.props" />
```
Comment (nguerrera) [inline src/NuGet/Microsoft.Net.Compilers.Toolset/buildMultiTargeting/Microsoft.Net.Compilers.Toolset.props]: Why is this file needed? Compilation happens in the inner build.

Comment (jaredpar) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/buildMultiTargeting/Microsoft.Net.Compilers.Toolset.props]: I'm following the guidelines here: https://natemcmaster.com/blog/2017/07/05/msbuild-task-in-nuget/ If those are incorrect I can remove. As is though it appears to be working based off the success of the bootstrap phase.

@natemcmaster

Comment (natemcmaster) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/buildMultiTargeting/Microsoft.Net.Compilers.Toolset.props]: If Nick is right and there is no task or target needed in the outer build of a multitargeting project, then this file can be removed. 

Related note: NuGet just added support for another folder, buildTransitive/. If you want  this package's targets to be imported when a project has a transitive reference to this package, add a file just like this one to that folder.

Comment (jaredpar) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/buildMultiTargeting/Microsoft.Net.Compilers.Toolset.props]: I think we want to avoid the transitive on for now. One "feature" of this package is you can install it on a single project to temporarily work around a compiler bug. The isolation there to that project is intended. It's possible that's not actually needed anymore but for the moment intent is to keep it (want to minimize the impact of this change). 

> If Nick is right and there is no task or target needed in the outer build of a multitargeting project, then this file can be removed.

Can either of you elaborate on this a bit? I'm not sure why this would or wouldn't be needed. I trust you two are right here but would like to understand this a bit better if possible.

File: src/Setup/DevDivVsix/CompilersPackage/Microsoft.CodeAnalysis.Compilers.Setup.csproj (hunk 3 above)
Comment (jaredpar) [reply src/Setup/DevDivVsix/CompilersPackage/Microsoft.CodeAnalysis.Compilers.Setup.csproj]: Done.

File: Roslyn.sln (hunk 2 above)
Comment (jaredpar) [reply Roslyn.sln]: Yes / No: the version changes whenever you open the SLN file in a new version of Visual Studio and save it. Generally we don't pay much attention to it as fixing it just means it will change for the next person. I'll revert this though.

File: src/NuGet/Microsoft.Net.Compilers.Toolset/Microsoft.Net.Compilers.Toolset.Package.csproj (hunk 5 above)
Comment (jaredpar) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/Microsoft.Net.Compilers.Toolset.Package.csproj]: Nope. Copy Pasta error. Fixing.

**Comment by jaredpar** [comment]
Note: there are a number of actions that will happen once this is merged. Those are all tracked here #33695

File: src/NuGet/Microsoft.Net.Compilers.Toolset/buildMultiTargeting/Microsoft.Net.Compilers.Toolset.props (hunk 6 above)
Comment (nguerrera) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/buildMultiTargeting/Microsoft.Net.Compilers.Toolset.props]: There are two different builds / evaluations of the the project that we call "inner build" and "outer build". In the outer build, TargetFramework is unset and build loops over the TargetFrameworks and performs individual inner builds with the target framework set. 

build/PackageName.props|targets is imported into inner builds only
buildMulltitarget/PackageName.props|targets is imported into outer builds only

Some of the confusion here may have to do with the fact that you aware multi-targeting the task, which makes `buildMultitargeting` seem relevant. However, this folder is not about the multi-targeting of the tasks in the package, but the multi-targeting of the projects that reference the package.

There's no one universal approach to deciding what to put in build/ vs buildMultitargeting/, it depends on what the task does and how it wants to achieve it.

1. Same code in inner build as outer build (what you have here by having buildMultitargeting turn around and import from build/)
2. Different code in inner and outer build (using different files or what you have + some conditions to handle the two cases differently).
3. Some code in inner build, no code at all in outer build
4. Some code in outer build, no code at all in inner build

Some things can just ignore that the outer build exists as they are purely features of the inner build. Compilation is like that. There's no CoreCompile in the outer build, only in the inner build where we are actually producing the assembly for a single target framework. Other things do have to handle the outer build in some way. Pack is a good example of that: we need a target that produces a single .nupkg that aggregates all the inner build outputs.

File: Compilers.sln [hunk 7]
```diff
@@ -160,6 +160,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "CompilerBenchmarks", "src\T
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "BuildBoss", "src\Tools\BuildBoss\BuildBoss.csproj", "{8A02AFAF-F622-4E3E-9E1A-8CFDACC7C7E1}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Net.Compilers.Toolset.Package", "src\NuGet\Microsoft.Net.Compilers.Toolset\Microsoft.Net.Compilers.Toolset.Package.csproj", "{6D407402-CC4A-4125-9B00-C70562A636A5}"
```
Comment (333fred) [inline Compilers.sln]: Consider adding this to the solution filter as well.

**Comment by jaredpar** [comment]
@tmat PTAL

File: src/NuGet/Microsoft.Net.Compilers.Toolset/build/Microsoft.Net.Compilers.Toolset.props [hunk 8]
```diff
@@ -0,0 +1,17 @@
+﻿<!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
+<Project DefaultTargets="Build">
+
+  <PropertyGroup>
+    <_TargetDirectoryName Condition="'$(MSBuildRuntimeType)' == 'Core'">netcoreapp2.1</_TargetDirectoryName>   
+    <_TargetDirectoryName Condition="'$(MSBuildRuntimeType)' != 'Core'">net472</_TargetDirectoryName>   
+    <_TasksDirectory>$(MSBuildThisFileDirectory)..\tasks\$(_TargetDirectoryName)\</_TasksDirectory>
```
Comment (nguerrera) [inline src/NuGet/Microsoft.Net.Compilers.Toolset/build/Microsoft.Net.Compilers.Toolset.props]: >_TasksDirectory [](start = 5, length = 15)

I suggest _RoslynTasksDirectoy, _RoslynTargetDirectoryName here. These will be visible to everyone else and the names are super generic.

File: src/NuGet/Microsoft.Net.Compilers.Toolset/build/Microsoft.Net.Compilers.Toolset.props [hunk 9]
```diff
@@ -0,0 +1,17 @@
+﻿<!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
+<Project DefaultTargets="Build">
```
Comment (nguerrera) [inline src/NuGet/Microsoft.Net.Compilers.Toolset/build/Microsoft.Net.Compilers.Toolset.props]: >DefaultTargets="Build" [](start = 9, length = 22)

I'm not sure if you should be specifying this. It is redundant with [https://github.com/Microsoft/msbuild/blob/99e150416ed43660782db8ea88b2d12135aca855/src/Tasks/Microsoft.Common.targets#L17](https://github.com/Microsoft/msbuild/blob/99e150416ed43660782db8ea88b2d12135aca855/src/Tasks/Microsoft.Common.targets#L17)

File: src/NuGet/Microsoft.Net.Compilers.Toolset/build/Microsoft.Net.Compilers.Toolset.props (hunk 8 above)
Comment (jaredpar) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/build/Microsoft.Net.Compilers.Toolset.props]: The intent is to make them visible. See Bootstrap.targets where we use the name. There are already other Roslyn prefixed variables in the build. That's why I chose this one here.

File: src/NuGet/Microsoft.Net.Compilers.Toolset/build/Microsoft.Net.Compilers.Toolset.props (hunk 9 above)
Comment (jaredpar) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/build/Microsoft.Net.Compilers.Toolset.props]: Okay. It's another thing I grabbed from the template on the web tutorial. Will delete.

File: src/NuGet/Microsoft.Net.Compilers.Toolset/build/Microsoft.Net.Compilers.Toolset.props (hunk 8 above)
Comment (nguerrera) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/build/Microsoft.Net.Compilers.Toolset.props]: I'm referring to the _ ones that are not Roslyn prefixed. I would put _Roslyn (or anything to differentiate).

Comment (jaredpar) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/build/Microsoft.Net.Compilers.Toolset.props]: Oh I thought `_` was the way to differentiate. I though it was MSBuild convention for `private`.

Comment (nguerrera) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/build/Microsoft.Net.Compilers.Toolset.props]: It is, but it's just a convention, so I would be slightly more defensive. It also makes it easier to find things in the binlog if you don't use super generic names in product targets for these sorts of things.

Comment (jaredpar) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/build/Microsoft.Net.Compilers.Toolset.props]: Gotcha. Fixed.

**Comment by tannergooding** [comment]
@jaredpar, Just wondering, but why do we need a new package rather than just reusing `Microsoft.Net.Compilers`?

Seems like (especially if we are planning on getting rid of these), we could just make the original/most commonly used package light up with .NET Core support in future versions.

**Comment by jaredpar** [comment]
@tannergooding please read the linked issue where this was discussed.

**Comment by tannergooding** [comment]
@jaredpar, I might be missing it, but I don't see why we couldn't update one of the existing packages mentioned in https://github.com/dotnet/roslyn/issues/33219.

The closest I see is the comment about not wanting to add more complexity for something that is primarily used internally. But, that doesn't explain why a new package is needed, since you should be able to just make `Microsoft.Net.Compilers` be identical to this new `Microsoft.Net.Compilers.Toolset` package. Then, people using it don't need to change anything but drop support for `Microsoft.NETCore.Compilers`, which will be removed at some point in the future.

**Comment by tannergooding** [comment]
To possibly clarify. I'm not understanding why we need `Microsoft.Net.Compilers.Toolset` at all.

I would have expected we could just have `Microsoft.Net.Compilers` be the "single package" and we could delete `Microsoft.NETCore.Compilers` in the future.

**Comment by jaredpar** [comment]
@tannergooding 

> I might be missing it, but I don't see why we couldn't update one of the existing packages mentioned in #33219.

Because there are consumers who depend on the directory structure of the existing package. They download the package, copy the binaries out and repackage them. If we change the directory structure we break them. The new package + deprecation of the existing ones is the cleanest way to do the migration.
